src/core/preprocessor.js

```diff
@@ -0,0 +1,256 @@
+import { Debug } from './debug.js';
+
+// set to true to enable debug logging
+const enableLogging = false;
+
+const log = (...args) => {
+    if (enableLogging) {
+        Debug.log(...args);
+    }
+};
+
+const logAssert = (condition, keyword, src, index) => {
+    Debug.assert(condition, `Invalid [${keyword}]: ${src.substring(index, index + 100)}...`);
+};
+
+// accepted keywords
+const KEYWORD = /([ ]*)+#(ifn?def|if|endif|else|elif|define|undef)/g;
+
+// #define EXPRESSION
+const DEFINE = /define[ ]+([^\n]+)\r?(?:\n|$)/g;
+
+// #undef EXPRESSION
+const UNDEF = /undef[ ]+([^\n]+)\r?(?:\n|$)/g;
+
+// #ifdef/#ifndef SOMEDEFINE, #if EXPRESSION
+const IF = /(ifdef|ifndef|if)[ ]*([^\r\n]+)\r?\n/g;
+
+// #endif/#else or #elif EXPRESSION
+const ENDIF = /(endif|else|elif)([ ]+[^\r\n]+)?\r?(?:\n|$)/g;
+
+// identifier
+const IDENTIFIER = /([\w-]+)/;
+
+// [!]defined(EXPRESSION)
+const DEFINED = /(!|\s)?defined\(([\w-]+)\)/;
+
+// currently unsupported characters in the expression: | & < > = + -
+const INVALID = /[><=|&+-]/g;
+
+/**
+ * Pure static class implementing subset of C-style preprocessor.
+ * inspired by: https://github.com/dcodeIO/Preprocessor.js
+ *
+ * @ignore
+ */
+class Preprocessor {
+    /**
+     * Run c-like proprocessor on the source code, and resolves the code based on the defines and ifdefs
+     *
+     * @param {string} source - The source code to work on.
+     * @returns {string} Returns preprocessed source code.
+     */
+    static run(source) {
```

Comment:
Shouldn't this function return success or failure?

Reply:
it's little error validation taking place in this function, but I return error (null) and handle it on the other side correctly now (there was an unrelated bug in shader compilation returning incorrect type)